Web Inspector: Add experimental feature to enable aggressive limits on the length of lines we let CodeMirror process/format

patrickangle opened the pull request:
#### b79f1636634fa4631caf970168c2bbe763edb3b7
<pre>
Web Inspector: Add experimental feature to enable aggressive limits on the length of lines we let CodeMirror process/format
<a href="https://bugs.webkit.org/show_bug.cgi?id=251401">https://bugs.webkit.org/show_bug.cgi?id=251401</a>
rdar://104840214

Reviewed by Devin Rousso and Justin Michaud.

In select cases, we are finding that Web Inspector is effectively unusable on some sites with many source files with
very longs lines of source code. The highlighting of very longs lines of code accounts for upwards of 30 seconds of
delay for very long lines of code. While we investigate further improvements we can make to solve this problem, we
are adding an experimental setting that enforces very low line limits for highlight, as well as for determining the
hovered token. These two changes will unblock developers encountering this issue when they enable the added experimental
setting.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorTokenTrackingController.js:
(WI.CodeMirrorTokenTrackingController.prototype._updateHoveredTokenInfo):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):
* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor):

Canonical link: <a href="https://commits.webkit.org/259603@main">https://commits.webkit.org/259603@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/285e0712eb3c5853ea98917a73ce6beae31c98b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105404 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14472 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114659 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5410 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97708 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111162 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95085 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26720 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/7786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7906 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/13929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47624 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6622 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9697 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/WebInspectorUI/UserInterface/Base/Setting.js:
```diff
@@ -235,6 +235,7 @@ WI.settings = {
     experimentalAllowInspectingInspector: new WI.Setting("experimental-allow-inspecting-inspector", false),
     experimentalCSSSortPropertyNameAutocompletionByUsage: new WI.Setting("experimental-css-sort-property-name-autocompletion-by-usage", true),
     experimentalEnableNetworkEmulatedCondition: new WI.Setting("experimental-enable-network-emulated-condition", false),
+experimentalAggressivelyLimitSourceCodeHighlighting: new WI.Setting("engineering-aggresively-limit-source-code-highlighting", false),
```
nit: indent

justinmichaud reviewed (approved):
I am not really qualified to review WI code in general, but this change is very nice and LGTM.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
```diff
@@ -37,6 +37,7 @@ WI.TextEditor = class TextEditor extends WI.View
             matchBrackets: true,
             autoCloseBrackets: true,
             styleSelectedText: true,
+            maxHighlightLength: WI.settings.experimentalAggressivelyLimitSourceCodeHighlighting.value ? 120 : 10000,
```
should we put this inside `WI.CodeMirrorEditor` instead so that all `CodeMirror` get it (just like `WI.settings.showWhitespaceCharacters`, etc.)?

dcrousso replied:
also, should we have something like this too?
```
WI.settings.experimentalAggressivelyLimitSourceCodeHighlighting.addEventListener(WI.Setting.Event.Changed, function(event) {
    this._codeMirror.setOption("maxHighlightLength", WI.settings.experimentalAggressivelyLimitSourceCodeHighlighting.value);
}, this)
```

dcrousso replied:
according to <[https://codemirror.net/5/doc/manual.html]()> it seems like `CodeMirror/matchbrackets.js` might also have some options we should change with this too (e.g. `maxScanLines`, `maxScanLineLength`, and `maxHighlightLineLength`)?

patrickangle replied:
For what it's worth I arrived at the two spots I limited based on profiling Web Inspector while inspecting a specific page. This setting wasn't necessarily meant to provide hard limits across the board, but rather to limit specific spots that had an outsized impact on performance. I'm happy to set those options as well, but that is why the setting currently didn't change other options.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorTokenTrackingController.js:
```diff
@@ -337,6 +337,9 @@ WI.CodeMirrorTokenTrackingController = class CodeMirrorTokenTrackingController e
     {
         // Get the position in the text and the token at that position.
         var position = this._codeMirror.coordsChar(mouseCoords);
+        if (WI.settings.experimentalAggressivelyLimitSourceCodeHighlighting.value && position.ch > 120)
```
should we do this for inline breakpoint stuff too? (see `limitLocations` inside `WI.SourceCodeTextEditor.prototype._addBreakpointWidgetsForLine`)

i wonder if there are any other things like these that could also apply 🤔

NIT: i kinda feel like this should be named something more general since it's being used for more than just highlighting :/

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -414,6 +414,11 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
             experimentalSettingsView.addSeparator();
         }
 
+        let sourcesGroup = experimentalSettingsView.addGroup(WI.unlocalizedString("Sources:"));
```
experimental settings should still be localized

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -414,6 +414,11 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
         }
 
+        let sourcesGroup = experimentalSettingsView.addGroup(WI.unlocalizedString("Sources:"));
+        sourcesGroup.addSetting(WI.settings.experimentalAggressivelyLimitSourceCodeHighlighting, WI.unlocalizedString("Aggresively limit source code line length limit for highlighting"));
```
NIT: I dont think "Aggressively" really adds much here beyond what "Limit" would already imply.  how about something simpler like "Limit syntax highlighting on long lines of code"?

patrickangle committed:
Web Inspector: Add experimental feature to enable aggressive limits on the length of lines we let CodeMirror process/format
https://bugs.webkit.org/show_bug.cgi?id=251401
rdar://104840214

Reviewed by Devin Rousso and Justin Michaud.

In select cases, we are finding that Web Inspector is effectively unusable on some sites with many source files with
very longs lines of source code. The highlighting of very longs lines of code accounts for upwards of 30 seconds of
delay for very long lines of code. While we investigate further improvements we can make to solve this problem, we
are adding an experimental setting that enforces very low line limits for highlight, as well as for determining the
hovered token. These two changes will unblock developers encountering this issue when they enable the added experimental
setting.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorTokenTrackingController.js:
(WI.CodeMirrorTokenTrackingController.prototype._updateHoveredTokenInfo):
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):
* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor):

Canonical link: https://commits.webkit.org/259603@main
